[#92] Support more error results in PrimingRequest

tolbertam opened the pull request:
Adds support for all protocol V2 error codes.  This is a work in progress as I moved some files between packages (was getting some namespace pollution) that probably needs revisiting and I may be missing a few tests, but this should be functional.

tolbertam commented on java-it-tests/common/src/main/java/common/CassandraResult.java:
```diff
@@ -43,6 +43,19 @@ public SuccessStatus() {
         }
     }
 
+    public static class ErrorMessageStatus extends ErrorStatus {
```
Did this in favor of creating a `ErrorStatus` for each response time as they only return error messages from the exception, though some exceptions may surface specific things, so I may add more to this (need to look).

tolbertam commented on java-it-tests/common/src/main/java/queries/QueryErrorPrimingTest.java:
```diff
@@ -84,4 +82,114 @@ public void testPrimingUnavailable() {
         assertEquals(3, ((CassandraResult.UnavailableStatus) status).getAlive());
     }
+
+    @Test
```
Couldn't find a clever way to share these tests between tests like I did with ErrorHandlingBehaviors using scalatest ((yay scalatest and traits!), but will see if I can find a way that doesn't involve sharing a parent class.

tolbertam commented on server/src/main/scala/org/scassandra/server/priming/json/ResultJsonRepresentation.scala:
```diff
@@ -0,0 +1,79 @@
+/*
```
I moved this into it's own json subpackage as some of the class names were clashing, but wasn't sure if this was a good idea, what do you think?

tolbertam replied:
If we go this route there's a few other json/webapi specific classes that'd be worth moving.

tolbertam replied:
Are you ok with the package change?  Only other thing I see that could be worth moving is PrimingJsonImplicits.   Can move that if it makes sense.

chbatey replied:
yep - I have also been meaning to move all the common things between PrimingJsonImplicits and the one for test into a separate class to remove the duplication.

tolbertam replied:
Sounds good, I'll move PrimingJsonImplicits and merge, thanks!

chbatey commented on java-client/src/main/java/org/scassandra/http/client/PrimingRequest.java:
```diff
@@ -357,7 +357,19 @@ public int hashCode() {
         read_request_timeout,
         unavailable,
-        write_request_timeout
+        write_request_timeout,
```
Who knew there were so many :)

chbatey commented on java-client/src/main/java/org/scassandra/http/client/UnpreparedConfig.java:
```diff
@@ -0,0 +1,32 @@
+import com.google.common.collect.ImmutableMap;
+import org.scassandra.server.priming.ErrorConstants;
+
+public class UnpreparedConfig extends Config {
```
Hadn't seen this error, perhaps I should use it rather than throw an error when scassandra actually hasn't got a prepare.

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/CqlMessageFactory.scala:
```diff
@@ -39,4 +51,37 @@ trait CqlMessageFactory {
 
   def parseQueryRequest(stream: Byte, byteString: ByteString, variableTypes: List[ColumnType[_]]): QueryRequest
+
+  def createErrorMessage(result: ErrorResult, stream: Byte, consistency: Consistency): Response = result match {
```
Much nicer to keep this in one place

chbatey commented on server/src/main/scala/org/scassandra/server/priming/routes/PrimeQueryResultExtractor.scala:
```diff
@@ -89,6 +130,18 @@ object PrimeQueryResultExtractor extends LazyLogging {
         case SuccessResult => Success
         case r: ReadRequestTimeoutResult => ReadTimeout
         case r: WriteRequestTimeoutResult => WriteTimeout
+        case r: ServerErrorResult => ServerError
```
we can probably use _ here, my bad

tolbertam replied:
good point, i'll fix.

chbatey commented on server/src/test/scala/org/scassandra/server/JavaDriverIntegrationTest.scala:
```diff
@@ -59,33 +61,84 @@ class JavaDriverIntegrationTest extends AbstractIntegrationTest with ScalaFuture
+      val whenQuery = "read timeout query"
+      prime(When(query = Some(whenQuery)), Then(None, result = Some(result), config = config))
+
+      intercept[T] {
```
Nice

chbatey commented on server/src/test/scala/org/scassandra/server/actors/ErrorHandlingBehaviors.scala:
```diff
@@ -0,0 +1,86 @@
+import org.scassandra.server.priming._
+
+trait ErrorHandlingBehaviors extends ProtocolProvider {
+  this: FunSuite =>
```
What does this statement do? Pardon my Scala ignorance :)

tolbertam replied:
Its a bit confusing, but its a way of saying that any type that has the ErrorHandlingBehaviors trait must also have the FunSuite trait but that ErrorHandlingBehaviors does not necessarily extend FunSuite.

tolbertam commented on server/src/test/scala/org/scassandra/server/JavaDriverIntegrationTest.scala:
```diff
@@ -59,33 +61,84 @@ class JavaDriverIntegrationTest extends AbstractIntegrationTest with ScalaFuture
+    expectException[InvalidQueryException](Invalid)
+  }
+
+  ignore("Test config error on query") {
```
Will be fixed in driver 2.0.11/2.1.8/2.2.2.  Will unignore once fixed.

tolbertam commented:
Fixed test issues and rebased on master, only thing that remains (I think) is the package moving in ResultJsonRepresentation and whether or not that is the right path (added another comment)

chbatey commented:
Feel free to merge this when you are ready

tolbertam commented:
@chbatey, I rebased, took care of the package move (051f6ac) and also updated ExecuteHandler to send back an 'Unprepared' error if a statement wasn't previously Prepared for a query being executed (as discussed [here](https://github.com/scassandra/scassandra-server/pull/97#discussion_r36256762)).  I didn't take care of that for BatchStatements yet, but figure I can do that in a separate PR.  Thought I'd ask you to take another quick look at the last 2 commits to make sure if everything is ok, and if all is wellt his is ready to merge.
